### PR TITLE
Make the batch processor limit data points rather than metrics.

### PR DESCRIPTION
**Pull request by kisieland**
**Description:**
Changes batch processor logic to limit data points rather than metrics.

**Link to tracking Issue:** #2754

**Testing:**
Ran `go test ./...`
I've added some edge case test and alerted existing tests.

### CODE REVIEW

**Comment by kisieland** [comment]
@dashpole @bogdandrutu can you take a look?

File: processor/batchprocessor/splitmetrics.go [hunk 1]
```diff
@@ -37,28 +39,29 @@ func splitMetrics(size int, src pdata.Metrics) pdata.Metrics {
...
+				totalCopiedDataPoints += srcDataPointCount
 				srcIlm.Metrics().MoveAndAppendTo(destIlm.Metrics())
 				return true
 			}
 
 			srcIlm.Metrics().RemoveIf(func(srcMetric pdata.Metric) bool {
 				// If we are done skip everything else.
-				if totalCopiedMetrics == size {
+				if srcMetric.DataPointCount()+totalCopiedDataPoints > size {
+					shouldSend = true
```
Comment (dashpole) [inline processor/batchprocessor/splitmetrics.go:60]: why do we add shouldSend here instead of just checking totalCopiedDataPoints similar to what we were doing before?

**Comment by dashpole** [comment]
This was approved during the 3/24 sig meeting: https://github.com/open-telemetry/opentelemetry-collector/issues/2754#issuecomment-805983282

**Comment by kisieland** [comment]
> I think this needs a release note in the CHANGELOG.md under breaking changes

Done

File: CHANGELOG.md [hunk 2]
```diff
@@ -671,8 +672,8 @@ Released 2020-06-16
 
 ## 🛑 Breaking changes 🛑
 
-- `isEnabled` configuration option removed (#909) 
-- `thrift_tchannel` protocol moved from `jaeger` receiver to `jaeger_legacy` in contrib (#636) 
+- `isEnabled` configuration option removed (#909)
+- `thrift_tchannel` protocol moved from `jaeger` receiver to `jaeger_legacy` in contrib (#636)
```
Comment (bogdandrutu) [inline CHANGELOG.md:677]: If you want to fix unnecessary whitespaces I am fine, but please do it in a separate PR.

File: consumer/pdata/metric.go [hunk 3]
```diff
@@ -127,31 +127,41 @@ func (md Metrics) MetricAndDataPointCount() (metricCount int, dataPointCount int
...
+		dataPointCount = ms.DoubleSum().DataPoints().Len()
+	case MetricDataTypeIntHistogram:
+		dataPointCount = ms.IntHistogram().DataPoints().Len()
+	case MetricDataTypeHistogram:
+		dataPointCount = ms.Histogram().DataPoints().Len()
+	case MetricDataTypeSummary:
+		dataPointCount = ms.Summary().DataPoints().Len()
+	}
+	return
+}
```
Comment (bogdandrutu) [inline consumer/pdata/metric.go:163]: Please avoid exposing public APIs unless strictly necessary and discussed in advance.

File: processor/batchprocessor/README.md [hunk 4]
```diff
@@ -15,8 +15,8 @@ any data drops such as sampling.
 Please refer to [config.go](./config.go) for the config spec.
 
 The following configuration options can be modified:
-- `send_batch_size` (default = 8192): Number of spans or metrics after which a
-batch will be sent regardless of the timeout.
+- `send_batch_size` (default = 8192): Number of spans or data point in metrics
+after which a batch will be sent regardless of the timeout.
```
Comment (bogdandrutu) [inline processor/batchprocessor/README.md:19]: Do we care that `batch` is one of the most used processor? Maybe have an option "split_by_metrics_dp" bool that defaults to false and if enable the limit is used for DP?

**Review by bogdandrutu — DISMISSED** [review]

File: processor/batchprocessor/README.md (hunk 4 above)
Comment (kisieland) [reply processor/batchprocessor/README.md:19]: That was my original proposition in #2754, but in the [notes](https://github.com/open-telemetry/opentelemetry-collector/issues/2754#issuecomment-805983282) from the meeting @dashpole says we should change the behaviour altogether. 
Should I get back to introducing the flag?

File: CHANGELOG.md (hunk 2 above)
Comment (kisieland) [reply CHANGELOG.md:677]: ack

File: consumer/pdata/metric.go (hunk 3 above)
Comment (kisieland) [reply consumer/pdata/metric.go:163]: I need this logic to split correctly.
The only other solution I can see is to move this logic to `splitmetrics.go` is this ok?

Comment (kisieland) [reply consumer/pdata/metric.go]: implemented this proposition. PTAL

File: processor/batchprocessor/README.md (hunk 4 above)
Comment (bogdandrutu) [reply processor/batchprocessor/README.md:19]: Not sure, @dashpole @tigrannajaryan @jrcamp what do you think?

Also `spans, metric data points, or log records... `

**Comment by tigrannajaryan** [comment]
Is there any measurable performance impact due to this change?

File: processor/batchprocessor/README.md (hunk 4 above)
Comment (dashpole) [reply processor/batchprocessor/README.md:19]: Based on what I remember from the discussion, we agreed that it wasn't very useful to batch by number of metrics, as that can give a varying amount of data in each batch.  We couldn't think of a reason why someone would want to batch by number of metrics over number of data points.

IMO if we think there _is_ a use-case for batching by metrics, I would still expect the default to be to batch by the number of data points.

**Comment by kisieland** [comment]
@tigrannajaryan I've added the benchmark feel free to run it.
On `main` branch I got:
```
$ GOMAXPROCS=12 go test -bench=BenchmarkBatchMetricProcessor -benchtime=200000x
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/collector/processor/batchprocessor
cpu: Intel(R) Xeon(R) CPU @ 2.30GHz
BenchmarkBatchMetricProcessor-12          200000            540101 ns/op
PASS
ok      go.opentelemetry.io/collector/processor/batchprocessor  110.234s
```

On `change-batch-processing` I've got:
```
$ GOMAXPROCS=12 go test -bench=BenchmarkBatchMetricProcessor -benchtime=200000x
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/collector/processor/batchprocessor
cpu: Intel(R) Xeon(R) CPU @ 2.30GHz
BenchmarkBatchMetricProcessor-12          200000            543281 ns/op
PASS
ok      go.opentelemetry.io/collector/processor/batchprocessor  110.903s
```

File: processor/batchprocessor/README.md [hunk 5]
```diff
@@ -15,8 +15,8 @@ any data drops such as sampling.
 Please refer to [config.go](./config.go) for the config spec.
 
 The following configuration options can be modified:
-- `send_batch_size` (default = 8192): Number of spans or metrics after which a
-batch will be sent regardless of the timeout.
+- `send_batch_size` (default = 8192): Number of spans, metric data points, or log
+records in metrics after which a batch will be sent regardless of the timeout.
```
Comment (bogdandrutu) [inline processor/batchprocessor/README.md:19]: Please document the split logic, since it is not a trivial split of data-points, actually you split at metric level but ensure data points max (which I think is the right thing to do).

Comment (kisieland) [reply processor/batchprocessor/README.md:19]: Done, wdyt about this version?

Comment (kisieland) [reply processor/batchprocessor/README.md:19]: @bogdandrutu friendly ping

File: processor/batchprocessor/README.md (hunk 4 above)
Comment (dashpole) [reply processor/batchprocessor/README.md]: I raised this at the sig meeting today, and we were still in agreement that we should change this behavior to batch by number of data points instead of by number of metrics.

**Comment by dashpole** [comment]
@jrcamp @Aneurysm9 It would be helpful to have your reviews as well, if you have time.

File: processor/batchprocessor/batch_processor.go [hunk 6]
```diff
@@ -274,19 +274,19 @@ func newBatchMetrics(nextConsumer consumer.Metrics) *batchMetrics {
 
 func (bm *batchMetrics) export(ctx context.Context, sendBatchMaxSize int) error {
 	var req pdata.Metrics
-	if sendBatchMaxSize > 0 && bm.metricCount > sendBatchMaxSize {
+	if sendBatchMaxSize > 0 && bm.dataPointCount > sendBatchMaxSize {
 		req = splitMetrics(sendBatchMaxSize, bm.metricData)
-		bm.metricCount -= sendBatchMaxSize
+		bm.dataPointCount -= sendBatchMaxSize
```
Comment (Aneurysm9) [inline processor/batchprocessor/batch_processor.go:279]: Is this correct?  The README seems to indicate that a split batch may contain less than the maximum number of data points if the next metric would put the batch over the cap.  I'd expect `bm.dataPointCount` to be reduced here by the number of data points in `req`.

File: processor/batchprocessor/batch_processor_test.go [hunk 7]
```diff
@@ -320,6 +322,8 @@ func TestBatchMetricProcessor_BatchSize(t *testing.T) {
 
 	requestCount := 100
 	metricsPerRequest := 5
+	dataPointsPreMetric := 2 // Since the int counter uses two datapoints.
+	dataPointsPerRequest := metricsPerRequest * dataPointsPreMetric
```
Comment (Aneurysm9) [inline processor/batchprocessor/batch_processor_test.go:326]:
```suggestion
	dataPointsPerMetric := 2 // Since the int counter uses two datapoints.
	dataPointsPerRequest := metricsPerRequest * dataPointsPerMetric
```

File: processor/batchprocessor/batch_processor_test.go [hunk 8]
```diff
@@ -357,7 +361,7 @@ func TestBatchMetricProcessor_BatchSize(t *testing.T) {
 	assert.Equal(t, 1, len(viewData))
 	distData := viewData[0].Data.(*view.DistributionData)
 	assert.Equal(t, int64(expectedBatchesNum), distData.Count)
-	assert.Equal(t, sink.MetricsCount(), int(distData.Sum()))
+	assert.Equal(t, sink.MetricsCount()*dataPointsPreMetric, int(distData.Sum()))
```
Comment (Aneurysm9) [inline processor/batchprocessor/batch_processor_test.go:364]:
```suggestion
	assert.Equal(t, sink.MetricsCount()*dataPointsPerMetric, int(distData.Sum()))
```

File: processor/batchprocessor/README.md (hunk 4 above)
Comment (tigrannajaryan) [reply processor/batchprocessor/README.md]: I agree, for metrics data points makes the most sense as the unit batch operates on.

File: processor/batchprocessor/batch_processor.go (hunk 6 above)
Comment (kisieland) [reply processor/batchprocessor/batch_processor.go:279]: Great catch, that is something that could have become very hard to debug issue.
I've created a fix together with a test to verify that is working correctly.

**Review by carlosalberto — APPROVED** [review]

**Comment by kisieland** [comment]
@bogdandrutu @Aneurysm9 please take a look

**Review by Aneurysm9 — APPROVED** [review]

File: CHANGELOG.md [hunk 9]
```diff
@@ -5,6 +5,7 @@
 ## 🛑 Breaking changes 🛑
 
 - Remove unused logstest package (#3222)
+- Change `batchprocessor` logic to limit data points rather than metrics (#3141)
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply CHANGELOG.md:9]: I am good with that.

File: processor/batchprocessor/README.md [hunk 10]
```diff
@@ -15,8 +15,11 @@ any data drops such as sampling.
...
```
Comment (bogdandrutu) [inline processor/batchprocessor/README.md:19]: `records in metrics` they are not only in metrics. Same everywhere you refer to metrics

File: processor/batchprocessor/batch_processor.go (hunk 6 above)
Comment (bogdandrutu) [reply processor/batchprocessor/batch_processor.go]: Is it not the case that after `splitMetrics` the returned object always has `sendBatchMaxSize`?

Comment (kisieland) [reply processor/batchprocessor/batch_processor.go]: In this case no.
It boils down to the fact that some metrics have more than one data point and `splitMetrics` not splitting the last metric to fill the batch fully.

File: processor/batchprocessor/README.md (hunk 10 above)
Comment (kisieland) [reply processor/batchprocessor/README.md]: I agree that `records in metric` was a wrong description.
But in the rest of the README it refers only to batching metrics and the non-even split.

Comment (kisieland) [reply processor/batchprocessor/README.md]: @bogdandrutu friendly ping

File: processor/batchprocessor/batch_processor.go (hunk 6 above)
Comment (bogdandrutu) [reply processor/batchprocessor/batch_processor.go]: Can you add a test for when the num points for a metric is greater than `sendBatchMaxSize`. I think we should split the metric as well.

Comment (kisieland) [reply processor/batchprocessor/batch_processor.go:279]: Added the logic that splits the metric by data points and the test that you described.
Additionally I changed the Readme to reflect the logic change.
@bogdandrutu PTAL

File: processor/batchprocessor/README.md (hunk 10 above)
Comment (kisieland) [reply processor/batchprocessor/README.md]: removed this part of the Readme as it is no longer valid.

**Comment by tigrannajaryan** [comment]
Please resolve the merge conflicts.

**Comment by kisieland** [comment]
Merge conflicts resolved. 
@tigrannajaryan @bogdandrutu please take a look, I would appreciate if this PR was included in next release.

**Review by tigrannajaryan — APPROVED** [review]
LGTM, thanks @kisieland

**Comment by tigrannajaryan** [comment]
@bogdandrutu PTAL, I believe all your comments are addressed and we can merge.

**Comment by tigrannajaryan** [comment]
@kisieland please squash the commits and add a descriptive commit message. Please also include the results of before/after benchmarking in the commit message.

**Comment by kisieland** [comment]
@tigrannajaryan Done

**Comment by tigrannajaryan** [comment]
@kisieland please fix the changelog conflict one more time (sorry).

**Comment by kisieland** [comment]
@tigrannajaryan I don't see any conflicts, I resolved them two times today.

File: processor/batchprocessor/splitmetrics.go [hunk 11]
```diff
@@ -37,29 +38,31 @@ func splitMetrics(size int, src pdata.Metrics) pdata.Metrics {
...
-				if totalCopiedMetrics == size {
+				if totalCopiedDataPoints == size {
 					return false
 				}
-				srcMetric.CopyTo(destIlm.Metrics().AppendEmpty())
-				totalCopiedMetrics++
-				return true
+				// If the metric has more data points than free slots we should split it.
+				newMetric, remove := splitMetric(srcMetric, size-totalCopiedDataPoints)
+				newMetric.CopyTo(destIlm.Metrics().AppendEmpty())
```
Comment (bogdandrutu) [inline processor/batchprocessor/splitmetrics.go:63]: Instead of returning a new metric then copying to dest, better to pass destMetric to split:
`splitMetric(srcMetric, destIlm.Metrics().AppendEmpty(), size-totalCopiedDataPoints)`

**Review by bogdandrutu — APPROVED** [review]
Small optimization to avoid creating a new metric then copy it to destination, but directly copy metadata and points to the destination metric.

File: processor/batchprocessor/splitmetrics.go (hunk 11 above)
Comment (kisieland) [reply processor/batchprocessor/splitmetrics.go]: done

**Comment by kisieland** [comment]
@bogdandrutu Please merge if everything is ok (merging is blocked for me)

**Comment by tigrannajaryan** [comment]
@kisieland Thank you for the contribution and for the patience.
